server/registry/test/grpctest/server.go

```diff
@@ -0,0 +1,134 @@
+package grpctest
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"testing"
+
+	"github.com/apigee/registry/rpc"
+	"github.com/apigee/registry/server/registry"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+)
+
+// NewIfNoAddress will create a RegistryServer served by a
+// basic grpc.Server if env var APG_REGISTRY_ADDRESS is
+// not set, otherwise will simply return a dummy Server
+// assuming the tests will connect to a the remote address.
+func NewIfNoAddress(rc registry.Config) (*Server, error) {
+	addr := os.Getenv("APG_REGISTRY_ADDRESS")
+	if addr != "" {
+		return &Server{}, nil
+	}
+	return NewServer(registry.Config{})
+}
+
+// TestMain can delegate here in packages that wish to
+// use TestMain for tests using a RegistryServer. This
+// uses NewIfNoAddress() semantics.
+func TestMain(m *testing.M, rc registry.Config) {
+	l, err := NewIfNoAddress(rc)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer l.Close()
+	os.Exit(m.Run())
+}
+
+// NewServer creates a RegistryServer served by a basic grpc.Server.
+// If rc.Database and rc.DBConfig are blank, a RegistryServer using
+// sqlite3 on a tmpDir is automatically created.
+// APG_REGISTRY_ADDRESS, APG_REGISTRY_AUDIENCES, and APG_REGISTRY_INSECURE
+// env vars are set for the client to connect to the created grpc service.
+// Call Close() when done to close server and clean up tmpDir as needed.
+// Example:
+// func TestXXX(t *testing.T) {
+// 	 l, err := grpctest.NewServer(registry.Config{})
+// 	 if err != nil {
+// 		t.Fatal(err)
+// 	 }
+// 	 defer l.Close()
+//   ... run test here ...
+// }
+func NewServer(rc registry.Config) (*Server, error) {
+	gl := &Server{}
+	var err error
+	if rc.Database == "" {
+		rc.Database = "sqlite3"
+	}
+	if rc.Database == "sqlite3" && rc.DBConfig == "" {
+		f, err := ioutil.TempFile("", "registry.db.*")
+		if err != nil {
+			return nil, err
+		}
+		rc.DBConfig = f.Name()
+		gl.TmpDir = f.Name()
+	}
+
+	rs, err := registry.New(rc)
+	if err != nil {
+		return nil, err
+	}
+
+	gl.Listener, err = net.Listen("tcp", "localhost:0") // random port
+	if err != nil {
+		gl.Close()
+		return nil, err
+	}
+
+	gl.Server = grpc.NewServer()
+	reflection.Register(gl.Server)
+	rpc.RegisterRegistryServer(gl.Server, rs)
+	rpc.RegisterAdminServer(gl.Server, rs)
+
```

Comment:
This seems like a lot of code duplication.

Reply:
Hmm. Do you mean with this? https://github.com/apigee/registry/blob/main/cmd/registry-server/main.go#L144. We could do a little refactoring once we decide this is a good direction.

Reply:
That's right. That's not in `server/registry` because I wanted to be able to mix in other servers (like a search server) from other future locations (at one time there was a search subdirectory that is now in registry-experimental). But a helper somewhere that does common work seems appropriate if you do this.